File: pfe-scala/app/controllers/Auctions.scala
package controllers

import akka.NotUsed
import akka.stream.scaladsl.{Sink, Flow, Source}
import models.{AuctionRooms, Shop}
import play.api.mvc.{Controller, WebSocket}
import play.api.libs.concurrent.Execution.Implicits.defaultContext
import play.api.libs.EventSource
import play.api.libs.json._
import scala.concurrent.Future

class Auctions(shop: Shop, auctionRooms: AuctionRooms) extends Controller {

  type Bid = (String, Double)
  implicit val writesNotification = Writes[Bid] {
    case (string: String, double: Double) => Json.obj("name" -> string, "price" -> double)
  }

  def jsonFlow[A : Writes]: Flow[A, JsValue, NotUsed] =
    Flow[A].map(a => Json.toJson(a))

  val bidValidator = (__ \ "price").read[Double]

  def room(id: Long) = AuthenticatedAction.async { implicit request =>
    shop.get(id) map {
      case Some(item) => Ok(views.html.auctionRoom(item))
      case None => NotFound
    }
  }

  def bid(id: Long) = AuthenticatedAction(parse.json(bidValidator)) { implicit request =>
      auctionRooms.bid(id, request.username, request.body)
      Ok
  }

  def notifications(id: Long) = AuthenticatedAction.async { implicit request =>
    auctionRooms.notifications(id).map { case (currentState, notifications) =>
      val allNotifications = Source(currentState.to[collection.immutable.Seq]).concat(notifications)
      Ok.chunked(allNotifications.via(jsonFlow).via(EventSource.flow))
    }
  }

  def roomWs(id: Long) = AuthenticatedAction.async { implicit request =>
    shop.get(id) map {
      case Some(item) => Ok(views.html.auctionRoomWs(item))
      case None => NotFound
    }
  }

  def channel(id: Long) = WebSocket.acceptOrResult[JsValue, JsValue] { implicit request =>
    Authentication.authenticated(name => {
       auctionRooms.notifications(id).map { case (currentState, notifications) =>
         val bidsHandler = Sink.foreach[JsValue] { json =>
           for (bid <- json.validate(bidValidator)) {
             auctionRooms.bid(id, name, bid)
           }
         }
         val allNotifications = Source(currentState.to[collection.immutable.Seq]).concat(notifications).via(jsonFlow)
         Right(Flow.fromSinkAndSource(bidsHandler, allNotifications))
       }
     },
      Future.successful(Left(Forbidden))
    )
  }

}

File: pfe-java/app/controllers/Auctions.java
package controllers;

import com.fasterxml.jackson.databind.JsonNode;
import models.AuctionRooms;
import models.Item;
import play.libs.EventSource;
import play.libs.Json;
import play.mvc.BodyParser;
import play.mvc.Result;
import play.mvc.WebSocket;

import javax.inject.Inject;
import javax.inject.Singleton;

import static controllers.AuthenticatedAction.Authenticated;
import static controllers.DBAction.DB;
import static play.libs.EventSource.Event.event;

@Singleton
public class Auctions extends Controller {

    @Inject
    public Auctions(Service service) {
        super(service);
    }

    public static class CreateBid {
        public Double price;
        public CreateBid() {}
    }

    @Authenticated
    @DB
    public Result room(Long id) {
        Item item = service.shop().get(id);
        if (item != null) {
            return ok(views.html.auctionRoom.render(item));
        } else return notFound();
    }

    @Authenticated
    @BodyParser.Of(BodyParser.Json.class)
    public Result bid(Long id) {
        CreateBid bid = Json.fromJson(request().body().asJson(), CreateBid.class);
        service.auctionRooms.bid(id, AuthenticatedAction.getUsername(ctx()), bid.price);
        return ok();
    }

    @Authenticated
    public Result notifications(Long id) {
        return ok(EventSource.whenConnected(eventSource -> service.auctionRooms
                .subscribe(id, (bid) -> eventSource.send(event(Json.toJson(bid))))
                .onRedeem(stateAndSubscription -> {
                    for (AuctionRooms.Bid bid : stateAndSubscription._1) {
                        eventSource.send(event(Json.toJson(bid)));
                    }
                    eventSource.onDisconnected(stateAndSubscription._2::cancel);
        })));
    }

    @Authenticated
    @DB
    public Result roomWs(Long id) {
        Item item = service.shop.get(id);
        if (item != null) {
            return ok(views.html.auctionRoomWs.render(item));
        } else return notFound();
    }

    public WebSocket<JsonNode> channel(Long id) throws Throwable {
        return Authentication.authenticated(ctx(), username -> WebSocket.<JsonNode>whenReady((in, out) -> {
            in.onMessage(json -> {
                CreateBid bid = Json.fromJson(json, CreateBid.class);
                service.auctionRooms.bid(id, username, bid.price);
            });
            service.auctionRooms
                    .subscribe(id, bid -> out.write(Json.toJson(bid)))
                    .onRedeem(stateAndSubscription -> {
                        for (AuctionRooms.Bid bid : stateAndSubscription._1) {
                            out.write(Json.toJson(bid));
                        }
                        in.onClose(stateAndSubscription._2::cancel);
                    });
        }), () -> WebSocket.reject(forbidden()));
    }

}
